packages/next/server/image-optimizer.ts

```diff
@@ -473,15 +477,35 @@ async function writeToCacheDir(
   await promises.writeFile(filename, buffer)
 }
 
+function getFileNameWithExtension(
+  url: string,
+  contentType: string | null
+): string | void {
+  const [urlWithoutQueryParams] = url.split('?')
+  const fileNameWithExtension = urlWithoutQueryParams.split('/').pop()
+  if (!contentType || !fileNameWithExtension) {
+    return
+  }
+
+  const [fileName] = fileNameWithExtension.split('.')
+  const extension = getExtension(contentType)
+  return `${fileName}.${extension}`
```

Comment:
the file extension is based on the `content-type` header, because of the caching and optimizations we have. Also, there might be cases when someone will use external image URLs that won't have the filename with extension in the URL.